core/trie/key.go

```diff
@@ -136,20 +134,46 @@ func (k *Key) Truncate(length uint8) {
 	}
 }
 
-func (k *Key) RemoveLastBit() {
-	if k.len == 0 {
-		return
-	}
+func (k *Key) String() string {
+	return fmt.Sprintf("(%d) %s", k.len, hex.EncodeToString(k.bitset[:]))
+}
 
-	k.len--
+// Copy returns a deep copy of the key
+func (k *Key) Copy() Key {
+	newKey := Key{len: k.len}
+	copy(newKey.bitset[:], k.bitset[:])
+	return newKey
+}
 
-	unusedBytes := k.unusedBytes()
-	clear(unusedBytes)
+func (k *Key) Bytes() [32]byte {
+	var result [32]byte
+	copy(result[:], k.bitset[:])
+	return result
+}
 
-	// clear upper bits on the last used byte
-	inUseBytes := k.inUseBytes()
-	unusedBitsCount := 8 - (k.len % 8)
-	if unusedBitsCount != 8 && len(inUseBytes) > 0 {
-		inUseBytes[0] = (inUseBytes[0] << unusedBitsCount) >> unusedBitsCount
+// findCommonKey finds the set of common MSB bits in two key bitsets.
+func findCommonKey(longerKey, shorterKey *Key) (Key, bool) {
+	divergentBit := findDivergentBit(longerKey, shorterKey)
+	commonKey := *shorterKey
+	commonKey.shiftRight(shorterKey.Len() - divergentBit + 1)
+	return commonKey, divergentBit == shorterKey.Len()+1
+}
+
+func findDivergentBit(longerKey, shorterKey *Key) uint8 {
+	divergentBit := uint8(0)
+	for divergentBit <= shorterKey.Len() &&
+		longerKey.IsBitSet(longerKey.Len()-divergentBit) == shorterKey.IsBitSet(shorterKey.Len()-divergentBit) {
+		divergentBit++
 	}
+	return divergentBit
+}
```

Comment:
Do you think we should add a check that the length of shorterKey is less than longerKey? Currently I think this might panic if they are inserted in the incorrect order

Reply:
If we look at the current usage of `findDivergentBit`, longerKey will always be of length `251`, and the max shorterKey can go is also `251`. So I don't think we need to add this check.